SharedPackages/VPN/Tests/VPNTests/WireGuardAdapterTests.swift

```diff
@@ -0,0 +1,552 @@
+        } else {
+            XCTFail("Expected attempt failure event")
+        }
+
+        pathMonitor.emitStatus(.satisfied)
+        XCTAssertEqual(eventHandler.handledEvents.count, 2)
+        if case .endTemporaryShutdownStateRecoveryFailure(let error) = eventHandler.handledEvents[1] {
+            guard let adapterError = error as? WireGuardAdapterError,
+                  case .setNetworkSettings(let underlyingError) = adapterError else {
+                XCTFail("Expected WireGuardAdapterError.setNetworkSettings, got \(error)")
+                return
+            }
+            XCTAssertEqual(underlyingError as? TestError, .someError)
+        } else {
+            XCTFail("Expected recovery failure event")
+        }
+
+        packetTunnelProvider.setTunnelNetworkSettingsError = nil
+        pathMonitor.emitStatus(.satisfied)
+        XCTAssertEqual(eventHandler.handledEvents.count, 3)
+        if case .endTemporaryShutdownStateRecoverySuccess = eventHandler.handledEvents[2] {
+            // success
+        } else {
+            XCTFail("Expected recovery success event")
+        }
+
+        XCTAssertEqual(wireGuardInterface.turnOnCallCount, 2)
+    }
+
+    func testUpdateWhileTemporaryShutdownDoesNotRestartBackend() {
+        startAdapterSuccessfully()
+        pathMonitor.emitStatus(.unsatisfied)
+
+        let expectation = expectation(description: "Update completes while temporary shutdown")
+        adapter.update(tunnelConfiguration: tunnelConfiguration, reassert: true) { error in
+            XCTAssertNil(error)
+            expectation.fulfill()
+        }
+        wait(for: [expectation], timeout: 10.0)
+
+        XCTAssertEqual(wireGuardInterface.setConfigCallCount, 0, "Backend should not receive config while offline")
+        XCTAssertEqual(wireGuardInterface.turnOnCallCount, 1, "Backend should not restart during update")
+    }
+    #elseif os(macOS)
+    func testMacPathUpdateBumpsSockets() {
+        startAdapterSuccessfully()
+        pathMonitor.emitStatus(.satisfied)
+        XCTAssertEqual(wireGuardInterface.bumpSocketsCallCount, 1)
+    }
+    #endif
+
+    private static func makePublicKey() -> PublicKey {
+        let hexKey = String(repeating: "ab", count: 32) // 32 bytes -> 64 hex characters
+        return PublicKey(hexKey: hexKey)!
+    }
+
+    @discardableResult
+    private func startAdapterSuccessfully(file: StaticString = #file, line: UInt = #line) -> XCTestExpectation {
+        let startExpectation = expectation(description: "Adapter starts")
+        adapter.start(tunnelConfiguration: tunnelConfiguration) { error in
+            XCTAssertNil(error, file: file, line: line)
+            startExpectation.fulfill()
+        }
+        wait(for: [startExpectation], timeout: 10.0)
+        return startExpectation
+    }
+
+}
+
+private enum TestError: Error, Equatable {
+    case someError
+}
+
+// MARK: - Mocks
+
+private final class MockWireGuardInterface: WireGuardGoInterface {
```

Comment:
I left this mock in this file since it's so specific to the adapter, but I'm going to move it out into its own file anyway for consistency - better to have it more widely accessible so that we can use it in packet tunnel provider tests anyway.